File: src/world.c
#include "world.h"

#include <math.h>
#include <string.h>

static float vec_dist(Vec2 a, Vec2 b)
{
    float dx = b.x - a.x;
    float dy = b.y - a.y;
    return sqrtf(dx * dx + dy * dy);
}

static void path_push(RoutePath *path, Vec2 point)
{
    if (path->count >= MAX_ROUTE_WAYPOINTS) {
        return;
    }
    if (path->count > 0) {
        path->path_length += vec_dist(path->waypoints[path->count - 1], point);
    }
    path->waypoints[path->count++] = point;
}

static void path_reset(RoutePath *path, LaneId lane, RouteType route)
{
    memset(path, 0, sizeof(*path));
    path->lane = lane;
    path->route = route;
}

static void build_paths_for_lane(World *world, LaneId lane)
{
    const float cx = CENTER_X;
    const float cy = CENTER_Y;
    const float h = INTERSECTION_HALF;
    const float lw = LANE_WIDTH;
    const float arm = ARM_LENGTH;

    const float n_sb_x = cx + lw * 0.5f;
    const float n_nb_x = cx - lw * 0.5f;
    const float s_nb_x = cx - lw * 0.5f;
    const float s_sb_x = cx + lw * 0.5f;
    const float e_wb_y = cy - lw * 0.5f;
    const float e_eb_y = cy + lw * 0.5f;
    const float w_eb_y = cy + lw * 0.5f;
    const float w_wb_y = cy - lw * 0.5f;

    RoutePath *left = &world->routes[lane][ROUTE_LEFT];
    RoutePath *right = &world->routes[lane][ROUTE_RIGHT];
    RoutePath *straight = &world->routes[lane][ROUTE_STRAIGHT];
    path_reset(left, lane, ROUTE_LEFT);
    path_reset(right, lane, ROUTE_RIGHT);
    path_reset(straight, lane, ROUTE_STRAIGHT);

    switch (lane) {
    case LANE_NORTH_SB:
        path_push(straight, (Vec2){n_sb_x, cy - h});
        path_push(straight, (Vec2){n_sb_x, cy + h});
        path_push(straight, (Vec2){n_sb_x, cy + h + arm});

        path_push(left, (Vec2){n_sb_x, cy - h});
        path_push(left, (Vec2){cx + h, cy - h});
        path_push(left, (Vec2){cx + h, e_eb_y});
        path_push(left, (Vec2){cx + h + arm, e_eb_y});

        path_push(right, (Vec2){n_sb_x, cy - h});
        path_push(right, (Vec2){cx - h, cy - h});
        path_push(right, (Vec2){cx - h, w_wb_y});
        path_push(right, (Vec2){cx - h - arm, w_wb_y});
        break;

    case LANE_SOUTH_NB:
        path_push(straight, (Vec2){s_nb_x, cy + h});
        path_push(straight, (Vec2){s_nb_x, cy - h});
        path_push(straight, (Vec2){s_nb_x, cy - h - arm});

        path_push(left, (Vec2){s_nb_x, cy + h});
        path_push(left, (Vec2){cx - h, cy + h});
        path_push(left, (Vec2){cx - h, w_wb_y});
        path_push(left, (Vec2){cx - h - arm, w_wb_y});

        path_push(right, (Vec2){s_nb_x, cy + h});
        path_push(right, (Vec2){cx + h, cy + h});
        path_push(right, (Vec2){cx + h, e_eb_y});
        path_push(right, (Vec2){cx + h + arm, e_eb_y});
        break;

    case LANE_WEST_EB:
        path_push(straight, (Vec2){cx - h, w_eb_y});
        path_push(straight, (Vec2){cx + h, w_eb_y});
        path_push(straight, (Vec2){cx + h + arm, w_eb_y});

        path_push(left, (Vec2){cx - h, w_eb_y});
        path_push(left, (Vec2){cx - h, cy - h});
        path_push(left, (Vec2){n_nb_x, cy - h});
        path_push(left, (Vec2){n_nb_x, cy - h - arm});

        path_push(right, (Vec2){cx - h, w_eb_y});
        path_push(right, (Vec2){cx - h, cy + h});
        path_push(right, (Vec2){s_sb_x, cy + h});
        path_push(right, (Vec2){s_sb_x, cy + h + arm});
        break;

    case LANE_EAST_WB:
        path_push(straight, (Vec2){cx + h, e_wb_y});
        path_push(straight, (Vec2){cx - h, e_wb_y});
        path_push(straight, (Vec2){cx - h - arm, e_wb_y});

        path_push(left, (Vec2){cx + h, e_wb_y});
        path_push(left, (Vec2){cx + h, cy + h});
        path_push(left, (Vec2){s_sb_x, cy + h});
        path_push(left, (Vec2){s_sb_x, cy + h + arm});

        path_push(right, (Vec2){cx + h, e_wb_y});
        path_push(right, (Vec2){cx + h, cy - h});
        path_push(right, (Vec2){n_nb_x, cy - h});
        path_push(right, (Vec2){n_nb_x, cy - h - arm});
        break;

    case LANE_NORTH_NB:
        path_push(straight, (Vec2){n_nb_x, cy - h - arm * 0.35f});
        path_push(straight, (Vec2){n_nb_x, cy - h - arm});
        path_push(left, (Vec2){n_nb_x, cy - h});
        path_push(left, (Vec2){cx - h, cy - h});
        path_push(left, (Vec2){cx - h, w_wb_y});
        path_push(right, (Vec2){n_nb_x, cy - h});
        path_push(right, (Vec2){cx + h, cy - h});
        path_push(right, (Vec2){cx + h, e_eb_y});
        break;

    case LANE_SOUTH_SB:
        path_push(straight, (Vec2){s_sb_x, cy + h + arm * 0.35f});
        path_push(straight, (Vec2){s_sb_x, cy + h + arm});
        path_push(left, (Vec2){s_sb_x, cy + h});
        path_push(left, (Vec2){cx + h, cy + h});
        path_push(left, (Vec2){cx + h, e_eb_y});
        path_push(right, (Vec2){s_sb_x, cy + h});
        path_push(right, (Vec2){cx - h, cy + h});
        path_push(right, (Vec2){cx - h, w_wb_y});
        break;

    case LANE_EAST_EB:
        path_push(straight, (Vec2){cx + h + arm * 0.35f, e_eb_y});
        path_push(straight, (Vec2){cx + h + arm, e_eb_y});
        path_push(left, (Vec2){cx + h, e_eb_y});
        path_push(left, (Vec2){cx + h, cy + h});
        path_push(left, (Vec2){s_sb_x, cy + h});
        path_push(right, (Vec2){cx + h, e_eb_y});
        path_push(right, (Vec2){cx + h, cy - h});
        path_push(right, (Vec2){n_nb_x, cy - h});
        break;

    case LANE_WEST_WB:
        path_push(straight, (Vec2){cx - h - arm * 0.35f, w_wb_y});
        path_push(straight, (Vec2){cx - h - arm, w_wb_y});
        path_push(left, (Vec2){cx - h, w_wb_y});
        path_push(left, (Vec2){cx - h, cy - h});
        path_push(left, (Vec2){n_nb_x, cy - h});
        path_push(right, (Vec2){cx - h, w_wb_y});
        path_push(right, (Vec2){cx - h, cy + h});
        path_push(right, (Vec2){s_sb_x, cy + h});
        break;

    default:
        break;
    }
}

static void setup_lane(World *world, LaneId id, const char *name, Vec2 spawn, Vec2 stop,
                       Vec2 light, float heading, int inbound)
{
    LaneInfo *lane = &world->lanes[id];
    lane->id = id;
    lane->name = name;
    lane->spawn = spawn;
    lane->stop_line = stop;
    lane->light_pos = light;
    lane->heading = heading;
    lane->inbound = inbound;
    lane->lane_length = vec_dist(spawn, stop);
}

void world_init(World *world)
{
    const float cx = CENTER_X;
    const float cy = CENTER_Y;
    const float h = INTERSECTION_HALF;
    const float lw = LANE_WIDTH;
    const float arm = ARM_LENGTH;

    memset(world, 0, sizeof(*world));

    setup_lane(world, LANE_NORTH_SB, "N_SB",
               (Vec2){cx + lw * 0.5f, cy - h - arm},
               (Vec2){cx + lw * 0.5f, cy - h},
               (Vec2){cx + lw * 0.5f, cy - h - 18.0f}, 90.0f, 1);

    setup_lane(world, LANE_NORTH_NB, "N_NB",
               (Vec2){cx - lw * 0.5f, cy - h - arm},
               (Vec2){cx - lw * 0.5f, cy - h},
               (Vec2){cx - lw * 0.5f, cy - h - 18.0f}, 270.0f, 0);

    setup_lane(world, LANE_SOUTH_NB, "S_NB",
               (Vec2){cx - lw * 0.5f, cy + h + arm},
               (Vec2){cx - lw * 0.5f, cy + h},
               (Vec2){cx - lw * 0.5f, cy + h + 18.0f}, 270.0f, 1);

    setup_lane(world, LANE_SOUTH_SB, "S_SB",
               (Vec2){cx + lw * 0.5f, cy + h + arm},
               (Vec2){cx + lw * 0.5f, cy + h},
               (Vec2){cx + lw * 0.5f, cy + h + 18.0f}, 90.0f, 0);

    setup_lane(world, LANE_EAST_WB, "E_WB",
               (Vec2){cx + h + arm, cy - lw * 0.5f},
               (Vec2){cx + h, cy - lw * 0.5f},
               (Vec2){cx + h + 18.0f, cy - lw * 0.5f}, 180.0f, 1);

    setup_lane(world, LANE_EAST_EB, "E_EB",
               (Vec2){cx + h + arm, cy + lw * 0.5f},
               (Vec2){cx + h, cy + lw * 0.5f},
               (Vec2){cx + h + 18.0f, cy + lw * 0.5f}, 0.0f, 0);

    setup_lane(world, LANE_WEST_EB, "W_EB",
               (Vec2){cx - h - arm, cy + lw * 0.5f},
               (Vec2){cx - h, cy + lw * 0.5f},
               (Vec2){cx - h - 18.0f, cy + lw * 0.5f}, 0.0f, 1);

    setup_lane(world, LANE_WEST_WB, "W_WB",
               (Vec2){cx - h - arm, cy - lw * 0.5f},
               (Vec2){cx - h, cy - lw * 0.5f},
               (Vec2){cx - h - 18.0f, cy - lw * 0.5f}, 180.0f, 0);

    for (int i = 0; i < LANE_COUNT; i++) {
        build_paths_for_lane(world, (LaneId)i);
    }
}

const LaneInfo *world_lane(const World *world, LaneId lane)
{
    if (lane < 0 || lane >= LANE_COUNT) {
        return NULL;
    }
    return &world->lanes[lane];
}

const RoutePath *world_route(const World *world, LaneId lane, RouteType route)
{
    if (lane < 0 || lane >= LANE_COUNT || route < 0 || route >= ROUTE_COUNT) {
        return NULL;
    }
    return &world->routes[lane][route];
}

LaneId world_lane_for_spawn_direction(int direction_index)
{
    static const LaneId map[4] = {
        LANE_SOUTH_NB,
        LANE_NORTH_SB,
        LANE_WEST_EB,
        LANE_EAST_WB,
    };
    if (direction_index < 0 || direction_index > 3) {
        return LANE_SOUTH_NB;
    }
    return map[direction_index];
}

int world_lane_capacity(const LaneInfo *lane)
{
    float slot = VEHICLE_LENGTH + SAFETY_GAP;
    if (slot <= 0.0f) {
        return 0;
    }
    return (int)floorf(lane->lane_length / slot);
}

File: src/world.rs
use crate::config;
use crate::types::{LaneId, RouteType, Vec2};

#[derive(Debug, Clone)]
pub struct LaneInfo {
    pub id: LaneId,
    pub spawn: Vec2,
    pub stop_line: Vec2,
    pub light_pos: Vec2,
    pub heading: f32,
    pub lane_length: f32,
    pub inbound: bool,
}

#[derive(Debug, Clone)]
pub struct RoutePath {
    pub lane: LaneId,
    pub route: RouteType,
    pub waypoints: Vec<Vec2>,
    pub path_length: f32,
}

#[derive(Debug, Clone)]
pub struct World {
    pub lanes: [LaneInfo; 8],
    pub routes: [[RoutePath; 3]; 8],
}

impl World {
    pub fn new() -> Self {
        let mut world = World {
            lanes: std::array::from_fn(|_| LaneInfo {
                id: LaneId::NorthSb,
                spawn: Vec2 { x: 0.0, y: 0.0 },
                stop_line: Vec2 { x: 0.0, y: 0.0 },
                light_pos: Vec2 { x: 0.0, y: 0.0 },
                heading: 0.0,
                lane_length: 0.0,
                inbound: false,
            }),
            routes: std::array::from_fn(|_| {
                std::array::from_fn(|_| RoutePath {
                    lane: LaneId::NorthSb,
                    route: RouteType::Straight,
                    waypoints: Vec::new(),
                    path_length: 0.0,
                })
            }),
        };
        init_lanes(&mut world);
        for lane in LaneId::ALL {
            build_paths_for_lane(&mut world, lane);
        }
        world
    }

    pub fn lane(&self, id: LaneId) -> &LaneInfo {
        &self.lanes[lane_index(id)]
    }

    pub fn route(&self, lane: LaneId, route: RouteType) -> &RoutePath {
        &self.routes[lane_index(lane)][route_index(route)]
    }
}

fn lane_index(id: LaneId) -> usize {
    match id {
        LaneId::NorthSb => 0,
        LaneId::NorthNb => 1,
        LaneId::SouthNb => 2,
        LaneId::SouthSb => 3,
        LaneId::EastWb => 4,
        LaneId::EastEb => 5,
        LaneId::WestEb => 6,
        LaneId::WestWb => 7,
    }
}

fn route_index(route: RouteType) -> usize {
    match route {
        RouteType::Left => 0,
        RouteType::Right => 1,
        RouteType::Straight => 2,
    }
}

fn dist(a: Vec2, b: Vec2) -> f32 {
    let dx = b.x - a.x;
    let dy = b.y - a.y;
    (dx * dx + dy * dy).sqrt()
}

fn setup_lane(
    lanes: &mut [LaneInfo; 8],
    id: LaneId,
    spawn: Vec2,
    stop: Vec2,
    light: Vec2,
    heading: f32,
    inbound: bool,
) {
    let lane = &mut lanes[lane_index(id)];
    lane.id = id;
    lane.spawn = spawn;
    lane.stop_line = stop;
    lane.light_pos = light;
    lane.heading = heading;
    lane.inbound = inbound;
    lane.lane_length = dist(spawn, stop);
}

fn init_lanes(world: &mut World) {
    let arm = config::ARM_LENGTH;
    let ix0 = config::IX0;
    let ix1 = config::IX1;
    let iy0 = config::IY0;
    let iy1 = config::IY1;
    let er_nb = config::ENTER_NB_X;
    let er_sb = config::ENTER_SB_X;
    let er_eb = config::ENTER_EB_Y;
    let er_wb = config::ENTER_WB_Y;
    let ex_nb = config::EXIT_NB_X;
    let ex_sb = config::EXIT_SB_X;
    let ex_eb = config::EXIT_EB_Y;
    let ex_wb = config::EXIT_WB_Y;

    // Inbound lanes (spec diagram enter arrows).
    setup_lane(
        &mut world.lanes,
        LaneId::NorthSb,
        Vec2 {
            x: er_sb,
            y: iy0 - arm,
        },
        Vec2 { x: er_sb, y: iy0 },
        Vec2 {
            x: er_sb - config::TRAFFIC_LIGHT_SIDE_OFFSET,
            y: iy0 - config::TRAFFIC_LIGHT_STOP_OFFSET,
        },
        90.0,
        true,
    );
    setup_lane(
        &mut world.lanes,
        LaneId::SouthNb,
        Vec2 {
            x: er_nb,
            y: iy1 + arm,
        },
        Vec2 { x: er_nb, y: iy1 },
        Vec2 {
            x: er_nb + config::TRAFFIC_LIGHT_SIDE_OFFSET,
            y: iy1 + config::TRAFFIC_LIGHT_STOP_OFFSET,
        },
        270.0,
        true,
    );
    setup_lane(
        &mut world.lanes,
        LaneId::WestEb,
        Vec2 {
            x: ix0 - arm,
            y: er_eb,
        },
        Vec2 { x: ix0, y: er_eb },
        Vec2 {
            x: ix0 - config::TRAFFIC_LIGHT_STOP_OFFSET,
            y: er_eb + config::TRAFFIC_LIGHT_SIDE_OFFSET,
        },
        0.0,
        true,
    );
    setup_lane(
        &mut world.lanes,
        LaneId::EastWb,
        Vec2 {
            x: ix1 + arm,
            y: er_wb,
        },
        Vec2 { x: ix1, y: er_wb },
        Vec2 {
            x: ix1 + config::TRAFFIC_LIGHT_STOP_OFFSET,
            y: er_wb - config::TRAFFIC_LIGHT_SIDE_OFFSET,
        },
        180.0,
        true,
    );

    // Outbound lanes (spec diagram departures on the opposite track).
    setup_lane(
        &mut world.lanes,
        LaneId::NorthNb,
        Vec2 {
            x: ex_nb,
            y: iy0 - arm,
        },
        Vec2 { x: ex_nb, y: iy0 },
        Vec2 {
            x: ex_nb,
            y: iy0 - 18.0,
        },
        270.0,
        false,
    );
    setup_lane(
        &mut world.lanes,
        LaneId::SouthSb,
        Vec2 {
            x: ex_sb,
            y: iy1 + arm,
        },
        Vec2 { x: ex_sb, y: iy1 },
        Vec2 {
            x: ex_sb,
            y: iy1 + 18.0,
        },
        90.0,
        false,
    );
    setup_lane(
        &mut world.lanes,
        LaneId::EastEb,
        Vec2 {
            x: ix1 + arm,
            y: ex_eb,
        },
        Vec2 { x: ix1, y: ex_eb },
        Vec2 {
            x: ix1 + 18.0,
            y: ex_eb,
        },
        0.0,
        false,
    );
    setup_lane(
        &mut world.lanes,
        LaneId::WestWb,
        Vec2 {
            x: ix0 - arm,
            y: ex_wb,
        },
        Vec2 { x: ix0, y: ex_wb },
        Vec2 {
            x: ix0 - 18.0,
            y: ex_wb,
        },
        180.0,
        false,
    );
}

struct PathBuilder {
    lane: LaneId,
    route: RouteType,
    points: Vec<Vec2>,
    length: f32,
}

impl PathBuilder {
    fn new(lane: LaneId, route: RouteType) -> Self {
        Self {
            lane,
            route,
            points: Vec::new(),
            length: 0.0,
        }
    }

    fn push(&mut self, p: Vec2) {
        if let Some(last) = self.points.last() {
            self.length += dist(*last, p);
        }
        self.points.push(p);
    }

    fn finish(self) -> RoutePath {
        RoutePath {
            lane: self.lane,
            route: self.route,
            waypoints: self.points,
            path_length: self.length,
        }
    }
}

fn set_route(world: &mut World, lane: LaneId, route: RouteType, path: RoutePath) {
    world.routes[lane_index(lane)][route_index(route)] = path;
}

fn build_paths_for_lane(world: &mut World, lane: LaneId) {
    let arm = config::ARM_LENGTH;
    let cx = config::CENTER_X;
    let cy = config::CENTER_Y;
    let ix0 = config::IX0;
    let ix1 = config::IX1;
    let iy0 = config::IY0;
    let iy1 = config::IY1;
    let er_nb = config::ENTER_NB_X;
    let er_sb = config::ENTER_SB_X;
    let er_eb = config::ENTER_EB_Y;
    let er_wb = config::ENTER_WB_Y;
    let ex_nb = config::EXIT_NB_X;
    let ex_sb = config::EXIT_SB_X;
    let ex_eb = config::EXIT_EB_Y;
    let ex_wb = config::EXIT_WB_Y;
    let hub = Vec2 { x: cx, y: cy };

    let mut straight = PathBuilder::new(lane, RouteType::Straight);
    let mut left = PathBuilder::new(lane, RouteType::Left);
    let mut right = PathBuilder::new(lane, RouteType::Right);

    match lane {
        // From north, southbound ↓ on west lane.
        LaneId::NorthSb => {
            let stop = Vec2 { x: er_sb, y: iy0 };
            straight.push(stop);
            straight.push(Vec2 { x: er_sb, y: iy1 });
            straight.push(Vec2 {
                x: ex_sb,
                y: iy1 + arm,
            });
            // Left → east: into the box, through center, then east.
            left.push(stop);
            left.push(Vec2 { x: er_sb, y: cy });
            left.push(hub);
            left.push(Vec2 { x: cx, y: er_eb });
            left.push(Vec2 {
                x: ix1 + arm,
                y: ex_eb,
            });
            // Right → west: through center, then west.
            right.push(stop);
            right.push(Vec2 { x: er_sb, y: cy });
            right.push(hub);
            right.push(Vec2 { x: ix0, y: ex_wb });
            right.push(Vec2 {
                x: ix0 - arm,
                y: ex_wb,
            });
        }
        // From south, northbound ↑ on east lane.
        LaneId::SouthNb => {
            let stop = Vec2 { x: er_nb, y: iy1 };
            straight.push(stop);
            straight.push(Vec2 { x: er_nb, y: iy0 });
            straight.push(Vec2 {
                x: ex_nb,
                y: iy0 - arm,
            });
            left.push(stop);
            left.push(Vec2 { x: er_nb, y: cy });
            left.push(hub);
            left.push(Vec2 { x: ix0, y: ex_wb });
            left.push(Vec2 {
                x: ix0 - arm,
                y: ex_wb,
            });
            right.push(stop);
            right.push(Vec2 { x: er_nb, y: cy });
            right.push(hub);
            right.push(Vec2 { x: ix1, y: ex_eb });
            right.push(Vec2 {
                x: ix1 + arm,
                y: ex_eb,
            });
        }
        // From west, eastbound → on south row.
        LaneId::WestEb => {
            let stop = Vec2 { x: ix0, y: er_eb };
            straight.push(stop);
            straight.push(Vec2 { x: ix1, y: er_eb });
            straight.push(Vec2 {
                x: ix1 + arm,
                y: ex_eb,
            });
            left.push(stop);
            left.push(Vec2 { x: cx, y: er_eb });
            left.push(hub);
            left.push(Vec2 { x: er_nb, y: cy });
            left.push(Vec2 {
                x: ex_nb,
                y: iy0 - arm,
            });
            right.push(stop);
            right.push(Vec2 { x: cx, y: er_eb });
            right.push(hub);
            right.push(Vec2 { x: er_sb, y: cy });
            right.push(Vec2 {
                x: ex_sb,
                y: iy1 + arm,
            });
        }
        // From east, westbound ← on north row.
        LaneId::EastWb => {
            let stop = Vec2 { x: ix1, y: er_wb };
            straight.push(stop);
            straight.push(Vec2 { x: ix0, y: er_wb });
            straight.push(Vec2 {
                x: ix0 - arm,
                y: ex_wb,
            });
            left.push(stop);
            left.push(Vec2 { x: cx, y: er_wb });
            left.push(hub);
            left.push(Vec2 { x: er_sb, y: cy });
            left.push(Vec2 {
                x: ex_sb,
                y: iy1 + arm,
            });
            right.push(stop);
            right.push(Vec2 { x: cx, y: er_wb });
            right.push(hub);
            right.push(Vec2 { x: er_nb, y: cy });
            right.push(Vec2 {
                x: ex_nb,
                y: iy0 - arm,
            });
        }
        LaneId::NorthNb => {
            straight.push(Vec2 {
                x: ex_nb,
                y: iy0 - arm * 0.35,
            });
            straight.push(Vec2 {
                x: ex_nb,
                y: iy0 - arm,
            });
        }
        LaneId::SouthSb => {
            straight.push(Vec2 {
                x: ex_sb,
                y: iy1 + arm * 0.35,
            });
            straight.push(Vec2 {
                x: ex_sb,
                y: iy1 + arm,
            });
        }
        LaneId::EastEb => {
            straight.push(Vec2 {
                x: ix1 + arm * 0.35,
                y: ex_eb,
            });
            straight.push(Vec2 {
                x: ix1 + arm,
                y: ex_eb,
            });
        }
        LaneId::WestWb => {
            straight.push(Vec2 {
                x: ix0 - arm * 0.35,
                y: ex_wb,
            });
            straight.push(Vec2 {
                x: ix0 - arm,
                y: ex_wb,
            });
        }
    }

    set_route(world, lane, RouteType::Straight, straight.finish());
    set_route(world, lane, RouteType::Left, left.finish());
    set_route(world, lane, RouteType::Right, right.finish());
}

pub fn print_lane_table(world: &World) {
    println!("\n--- Lane data (Person 1, Rust) ---");
    println!("capacity = floor(lane_length / (vehicle_length + safety_gap))");
    println!(
        "vehicle_length={} safety_gap={} slot={}\n",
        config::VEHICLE_LENGTH,
        config::SAFETY_GAP,
        config::VEHICLE_LENGTH + config::SAFETY_GAP
    );
    for id in LaneId::ALL {
        let lane = world.lane(id);
        println!(
            "{:<8} spawn({:6.0},{:6.0}) stop({:6.0},{:6.0}) len={:6.1} cap={} inbound={}",
            id.name(),
            lane.spawn.x,
            lane.spawn.y,
            lane.stop_line.x,
            lane.stop_line.y,
            lane.lane_length,
            crate::types::lane_capacity(lane.lane_length),
            lane.inbound as u8
        );
    }
    println!("\nRoute colors: LEFT=blue RIGHT=yellow STRAIGHT=green");
    println!("Controls (Person 3): arrows spawn, r random, Esc quit");
    println!("Queue hook get_lane_queue_count() — live when vehicles are queued\n");
}
